concourse/steps/release.py

```diff
@@ -23,19 +28,303 @@
+        self.release_version = not_empty(release_version)
+        self.rebase_before_release = rebase_before_release
+        self.release_commit_callback = release_commit_callback
+
+    def name(self):
+        return "Release Commit Creator"
+
+    def validate(self):
+        info(f'validating { self.name()}')
+         # ensure the given callback script exists
+        if self.release_commit_callback:
+            self.release_commit_callback = os.path.join(
+                self.repo_dir,
+                self.release_commit_callback,
+            )
+            existing_file(self.release_commit_callback)
+
+        # ensure there is a file at the given version file path
+        existing_file(self.repository_version_file_path)
+
+    def apply(self):
+        if self.rebase_before_release:
+            upstream_commit_sha = self.git_helper.fetch_head(
+                f'refs/heads/{self.repository_branch}'
+            ).hexsha
+            self.git_helper.rebase(commit_ish=upstream_commit_sha)
+
+        # clean repository if required
+        worktree_dirty = bool(self.git_helper._changed_file_paths())
+        if worktree_dirty:
+            self.git_helper._stash_changes()
+
+        info('Creating release commit')
+        # update version file
+        self.repository_version_file_path.write_text(self.release_version)
+
+        try:
+            # call optional release commit callback
+            if self.release_commit_callback:
+                self._invoke_release_callback(
+                    release_commit_callback=self.release_commit_callback,
+                    repo_dir=self.repo_dir,
+                    release_version=self.release_version,
+                )
+
+            release_commit = self.git_helper.index_to_commit(
+                message=f'Release {self.release_version}',
+            )
+
+            release_commit_sha = release_commit.hexsha
+            self.release_context.set_release_commit_sha(release_commit_sha)
+
+            # forward head to new commit
+            self.git_helper.repo.head.set_commit(release_commit_sha)
+
+            info('Pushing release commit')
+
+            # Push release commit to remote
+            self.git_helper.push(
+                from_ref=release_commit_sha,
+                to_ref=self.repository_branch,
+                use_ssh=True,
+            )
+
+        finally:
+            if worktree_dirty and self.git_helper._has_stash():
+                self.git_helper._pop_stash()
+
+    def _invoke_release_callback(
+        self,
+        release_commit_callback,
+        repo_dir,
+        release_version,
+    ):
+        callback_env = os.environ.copy()
+        callback_env['REPO_DIR'] = repo_dir
+        callback_env['EFFECTIVE_VERSION'] = release_version
+
+        subprocess.run(
+            [release_commit_callback],
+            check=True,
+            env=callback_env,
+        )
+
+    def revert(self):
+        # TODO: revert commit
```

Comment:
how about omitting the unimplemented method and print a warning instead (in the base class)?
at any rate: do not return silently